server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateShardPathIT.java

```diff
@@ -63,12 +71,22 @@ public void testCheckShards() throws Exception {
         updateIndexSettings(indexName, Settings.builder().put("index.routing.allocation.exclude._name", node2));
         ensureGreen(indexName);
         assertBusy(() -> {
-            // The excluded node should eventually delete the shards
-            PrevalidateShardPathRequest req2 = new PrevalidateShardPathRequest(shardIdsToCheck, node2Id);
-            PrevalidateShardPathResponse resp2 = client().execute(TransportPrevalidateShardPathAction.TYPE, req2).get();
-            assertThat(resp2.getNodes().size(), equalTo(1));
-            assertTrue(resp.failures().isEmpty());
-            assertTrue(resp2.getNodes().get(0).getShardIds().isEmpty());
-        });
+            try {
+                // The excluded node should eventually delete the shards
+                PrevalidateShardPathRequest req2 = new PrevalidateShardPathRequest(shardIdsToCheck, node2Id);
+                PrevalidateShardPathResponse resp2 = client().execute(TransportPrevalidateShardPathAction.TYPE, req2).get();
+                assertThat(resp2.getNodes().size(), equalTo(1));
+                assertThat(resp2.getNodes().get(0).getNode().getId(), equalTo(node2Id));
+                assertTrue("There should be no failures in the response", resp.failures().isEmpty());
+                assertTrue("The relocation source node should have removed the shard(s)", resp2.getNodes().get(0).getShardIds().isEmpty());
+            } catch (AssertionError e) {
+                // Removal of shards which are no longer allocated to the node is attempted on every cluster state change in IndicesStore.
+                // If for whatever reason the removal is not triggered (e.g. not enough nodes reported that the shards are active) or it
+                // temporarily failed to clean up the shard folder, we need to trigger another cluster state change for this removal to
+                // finally succeed.
+                updateIndexSettings(indexName, Settings.builder().put("index.routing.allocation.exclude.name", "non-existent"));
```

Comment:
I think we need to add a random bit to the exclude:

```suggestion
                updateIndexSettings(indexName, Settings.builder().put("index.routing.allocation.exclude.name", "non-existent" + randomAlphaOfLength(5)));
```

since index settings that change nothing should not result in cluster state changes.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/92112/commits/8a511a156cdc8ed879ebaa60d591d7c24fc37cf1.